cmd/createorupdate/createorupdate.go

```diff
@@ -13,47 +16,147 @@ import (
 	"github.com/openshift/openshift-azure/pkg/fakerp"
 )
 
-func main() {
-	logger := logrus.New()
-	logger.Formatter = &logrus.TextFormatter{FullTimestamp: true}
-	logger.SetLevel(logrus.DebugLevel)
-	log := logrus.NewEntry(logger)
+type server struct {
+	// the server will not process more than a single
+	// request at all times.
+	inProgress chan struct{}
+	log        *logrus.Entry
+}
+
+// ServeHTTP handles an incoming request to the server.
+func (s *server) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	defer req.Body.Close()
 
-	// read in the external API manifest.
-	b, err := ioutil.ReadFile("_data/manifest.yaml")
+	ok := s.validate(w, req, nil)
+	if !ok {
+		return
+	}
+	defer func() {
+		// drain once we are done processing this request
+		<-s.inProgress
+	}()
+
+	b, err := ioutil.ReadAll(req.Body)
 	if err != nil {
-		log.Fatal(err)
+		resp := "500 Internal Server Error: Failed to read request body"
+		logrus.Debug(resp, err)
+		http.Error(w, resp, http.StatusInternalServerError)
+		return
 	}
+
 	var oc *v20180930preview.OpenShiftManagedCluster
-	err = yaml.Unmarshal(b, &oc)
-	if err != nil {
-		log.Fatal(err)
+	if err := yaml.Unmarshal(b, &oc); err != nil {
+		resp := "500 Internal Server Error: Failed to unmarshal request"
+		logrus.Debug(resp, err)
+		http.Error(w, resp, http.StatusInternalServerError)
+		return
 	}
 
-	//simulate Context with property bag
+	// simulate Context with property bag
 	ctx := context.Background()
 	ctx = context.WithValue(ctx, api.ContextKeyClientID, os.Getenv("AZURE_CLIENT_ID"))
 	ctx = context.WithValue(ctx, api.ContextKeyClientSecret, os.Getenv("AZURE_CLIENT_SECRET"))
 	ctx = context.WithValue(ctx, api.ContextKeyTenantID, os.Getenv("AZURE_TENANT_ID"))
 
-	// simulate the API call to the RP
-	entry := logrus.NewEntry(logger).WithFields(logrus.Fields{"resourceGroup": os.Getenv("RESOURCEGROUP")})
 	config := &api.PluginConfig{
 		SyncImage:       os.Getenv("SYNC_IMAGE"),
 		AcceptLanguages: []string{"en-us"},
 	}
-	oc, err = fakerp.CreateOrUpdate(ctx, oc, entry, config)
+
+	oc, err = fakerp.CreateOrUpdate(ctx, oc, s.log, config)
 	if err != nil {
-		log.Fatal(err)
+		resp := "500 Internal Server Error: Failed to apply request"
+		logrus.Debug(resp, err)
+		http.Error(w, resp, http.StatusInternalServerError)
+		return
 	}
 
-	// persist the returned (updated) external API manifest.
-	b, err = yaml.Marshal(oc)
+	res, err := yaml.Marshal(oc)
+	if err != nil {
+		resp := "500 Internal Server Error: Failed to marshal response"
+		logrus.Debug(resp, err)
+		http.Error(w, resp, http.StatusInternalServerError)
+		return
+	}
+	w.Write(res)
+}
+
+func (s *server) ListenAndServe() {
+	http.Handle("/", s)
+	httpServer := &http.Server{Addr: "localhost:8080"}
+	s.log.Info("starting server on localhost:8080")
+	logrus.WithError(httpServer.ListenAndServe()).Warn("Server exited.")
+}
+
+func (s *server) validate(w http.ResponseWriter, r *http.Request, expected []byte) bool {
+	// TODO: Support DELETE
```

Comment:
and also GET - currently we're implementing an asynchronous process synchronously, which is OK, but the real client would poll after the PUT completes